usr/share/rear/lib/authtoken-functions.sh

```diff
@@ -0,0 +1,546 @@
+#!/bin/bash
+
+function authtkn_load() {
+	# $1 container path
+	# $2 container offset
+	# $3 encryption key
+	# $4 [bound_to_blockdev path]
+
+	local intermediate_key=""
+	local cipher
+	local plain
+	local ret
+
+	cipher=$(token_read "$1" "$2"); ret=$?
+	if [ $ret -ne 0 ]; then
+		echo "authtkn_load(): token_read() failed: $ret" >&2
+		return $ret
+	fi
+
+	if [ -n "$4" ]; then
+		intermediate_key=$(blkdev_hash "$4" 2); ret=$?
+		if [ $ret -ne 0 ]; then
+			echo "authtkn_load(): failed to derive intermediate key from $4: $ret" >&2
+			return $ret
+		fi
+	fi
+
+	plain=$(decrypt_base64 "$cipher" "$3" "$intermediate_key"); ret=$?
+	if [ $ret -ne 0 ]; then
+		echo "authtkn_load(): decrypt_base64() failed: $ret" >&2
+		return $ret
+	fi
+
+	echo $plain
+	return 0
+}
+
+function authtkn_store() {
+	# $1 container path
+	# $2 container offset
+	# $3 token plaintext
+	# $4 encryption key
+	# $5 [bind_to_blockdev path]
+
+	local intermediate_key=""
+	local cipher
+	local tknlen
+	local ret
+
+	if [ -n "$5" ]; then
+		intermediate_key=$(blkdev_hash "$5" 2); ret=$?
+		if [ $ret -ne 0 ]; then
+			echo "authtkn_store(): failed to derive intermediate key from $5: $ret" >&2
+			return $ret
+		fi
+	fi
+
+	cipher=$(encrypt_base64 "$3" "$4" "$intermediate_key"); ret=$?
+	if [ $ret -ne 0 ]; then
```

Comment:
Don't test $? in a separated command
because that does not work reliably.
E.g. it falls apart when `set -e` is set.
For details see the comment about 'Using $?' in
usr/share/rear/lib/_input-output-functions.sh
https://github.com/rear/rear/blob/rear-2.7/usr/share/rear/lib/_input-output-functions.sh#L794

Reply:
I see now that in this particular case
one cannot use something like
```
if ! command_stdout="$( COMMAND )" ; then
    DebugPrint "COMMAND failed with $?"
fi
```
because
```
# if ! command_stdout="$( cat qqq )" ; then echo "$?" ; fi
cat: qqq: No such file or directory
0
```
but
```
# command_stdout="$( cat qqq )" ; echo "$?"
cat: qqq: No such file or directory
1
```

Reply:
But using PIPESTATUS seems to work
(at least for me with GNU bash version 4.4.23):
```
# if ! command_stdout="$( cat qqq )" ; then declare -p PIPESTATUS ; fi
cat: qqq: No such file or directory
declare -a PIPESTATUS=([0]="1")

# if ! command_stdout="$( cat qqq )" ; then echo "assignment failed with $PIPESTATUS" ; fi
cat: qqq: No such file or directory
assignment failed with 1

# if ! command_stdout="$( cat /etc/issue | grep qqq )" ; then echo "assignment failed with $PIPESTATUS" ; fi
assignment failed with 1

# if ! command_stdout="$( cat qqq | grep qqq )" ; then echo "assignment failed with $PIPESTATUS" ; fi
cat: qqq: No such file or directory
assignment failed with 1

# if ! command_stdout="$( cat /etc/issue | grep -Q qqq )" ; then echo "assignment failed with $PIPESTATUS" ; fi
grep: invalid option -- 'Q'
Usage: grep [OPTION]... PATTERN [FILE]...
Try 'grep --help' for more information.
assignment failed with 2

# if ! command_stdout="$( cat qqq | grep -Q qqq )" ; then echo "assignment failed with $PIPESTATUS" ; fi
cat: qqq: No such file or directory
grep: invalid option -- 'Q'
Usage: grep [OPTION]... PATTERN [FILE]...
Try 'grep --help' for more information.
assignment failed with 2
```

Because PIPESTATUS is an array that contains the exit status values
from the processes in the most recently executed foreground pipeline
PIPESTATUS is not from the pipe in the command substitution
but PIPESTATUS is from the single assignment command
```
# if ! command_stdout="$( cat qqq | grep -Q qqq )" ; then echo "assignment failed with $( declare -p PIPESTATUS )" ; fi
cat: qqq: No such file or directory
grep: invalid option -- 'Q'
Usage: grep [OPTION]... PATTERN [FILE]...
Try 'grep --help' for more information.
assignment failed with declare -a PIPESTATUS=([0]="2")
```
in contrast to the PIPESTATUS from the pipe commands
```
# cat qqq | grep -Q qqq ; declare -p PIPESTATUS
cat: qqq: No such file or directory
grep: invalid option -- 'Q'
Usage: grep [OPTION]... PATTERN [FILE]...
Try 'grep --help' for more information.
declare -a PIPESTATUS=([0]="1" [1]="2")
```

So this should work:
```
if ! command_stdout="$( COMMAND )" ; then
    DebugPrint "COMMAND failed with $PIPESTATUS"
fi
```

Reply:
@pcahyna 
could you have a look here about my finding that
(at least for me with GNU bash version 4.4.23)
```
if ! command_stdout="$( COMMAND )" ; then
    DebugPrint "COMMAND failed with $?"
fi
```
does not work because $=? is always 0
(i.e. $? does no longer hold the exit status of COMMAND)
but
```
if ! command_stdout="$( COMMAND )" ; then
    DebugPrint "COMMAND failed with $PIPESTATUS"
fi
```
does work because $PIPESTATUS (i.e. ${PIPESTATUS[0]})
keeps the exit status of COMMAND.

Is this behaviour perhaps even explained in "man bash"?
My GNU bash version 4.4.23 "man bash" shows (excerpts)
```
  ?   Expands to the exit status of the most recently
      executed foreground pipeline.
...
PIPESTATUS
      An array variable (see Arrays below) containing
      a list of exit status values from the processes
      in the most-recently-executed foreground pipeline
      (which may contain only a single command).
```

Reply:
Yes, `PIPESTATUS` is in almost all cases better than `$?`.

Reading this code area in ReaR I'm wondering if maybe communicating via return code is the core problem, and makes the code complicated and difficult to read.

Would it be possible to limit the use of return codes to either `0` for good or `>0` for bad and not care about the specific value of the return code? To convey more error details maybe it would be better to change the output of a function in case of error. It is also valid to `Error()` out *in* a function if there is no point to continue running ReaR at this moment, e.g. for fatal errors.

For example like this:
```shell
$ function err() { echo error message here ; return 1 ; }
$ if result=$(err) ; then echo "good $result" ; else echo "Error: $result" ; fi
Error: error message here
$ echo $result
error message here
```

Reply:
It is not related to assignments or command substitution:
```
# if ! cat qqq ; then echo "failed with $?" ; fi
cat: qqq: No such file or directory
failed with 0

# if ! cat qqq ; then echo "failed with $PIPESTATUS" ; fi
cat: qqq: No such file or directory
failed with 1
```

Reply:
Got it!
```
# true ; echo $?
0

# false ; echo $?
1

# ! true ; echo $?
1

# ! false ; echo $?
0

# true ; echo $PIPESTATUS
0

# false ; echo $PIPESTATUS
1

# ! true ; echo $PIPESTATUS
0

# ! false ; echo $PIPESTATUS
1

# if cat qqq ; then : ; else  echo "failed with $?" ; fi
cat: qqq: No such file or directory
failed with 1

# if command_stdout="$( cat /etc/issue | grep qqq )" ; then : ; else echo "assignment failed with $?" ; fi
assignment failed with 1

# if command_stdout="$( cat qqq | grep qqq )" ; then : ; else echo "assignment failed with $?" ; fi
cat: qqq: No such file or directory
assignment failed with 1

# if command_stdout="$( cat /etc/issue | grep -Q qqq )" ; then : ; else echo "assignment failed with $?" ; fi
grep: invalid option -- 'Q'
Usage: grep [OPTION]... PATTERN [FILE]...
Try 'grep --help' for more information.
assignment failed with 2

# if command_stdout="$( cat qqq | grep -Q qqq )" ; then : ; else echo "assignment failed with $?" ; fi
cat: qqq: No such file or directory
grep: invalid option -- 'Q'
Usage: grep [OPTION]... PATTERN [FILE]...
Try 'grep --help' for more information.
assignment failed with 2
```
This is because '!' is a reserverd word in bash
that behaves like a command i.e. it sets $?

Reply:
Buttom line:
Be careful with '!'.
Its name '!' tells all ;-)

Reply:
So the conclusion is to avoid `!` if you care about the actual numerical value of the error?

Thanks for the research! I didn't know this detail.

Maybe we can simplify the complexity of Bash and error code handling if we try to avoid using the numerical value of the return code, as much as feasible?

Reply:
It seems we don't have code in ReaR of the form
```
... if ! ...
    ...
    ... $? ...
```
i.e. code that uses '$?' shortly after 'if !'
because
```
# find usr/sbin/rear usr/share/rear -type f | xargs grep -A3 'if !' | grep '$?'
```
shows nothing.

Reply:
Yes, which is why I'd really prefer to not use the return code for anything beyond "is not 0"

Reply:
Thats why I assign $? to a variable here right after command in question.
However `$PIPESTATUS` might be a better approach, I'll look into it when time allows